collector/metadata/kubernetes/pod_watch.go

```diff
@@ -248,22 +240,115 @@ func OnUpdate(objOld interface{}, objNew interface{}) {
 		// Two different versions of the same pod will always have different RVs.
 		return
 	}
-	podUpdateMutex.Lock()
-	// TODO: re-implement the updated logic to reduce computation
-	onDelete(objOld)
+
+	oldCachePod, ok := globalPodInfo.get(oldPod.Namespace, oldPod.Name)
+	if !ok {
+		onAdd(objNew)
+		return
+	}
+	// Always override the old pod in the cache
 	onAdd(objNew)
-	podUpdateMutex.Unlock()
+
+	// Delay delete the pod using the difference between the old pod and the new one
+	deletedPodInfo := &deletedPodInfo{
+		name:         "",
+		namespace:    oldPod.Namespace,
+		containerIds: nil,
+		ip:           oldPod.Status.PodIP,
+		ports:        nil,
+		hostIp:       oldPod.Status.HostIP,
+		hostPorts:    nil,
+	}
+
+	if oldPod.Name != newPod.Name {
+		deletedPodInfo.name = oldPod.Name
+	}
+
+	// Check the containers' ID
+	newContainerIds := make([]string, 0)
+	for _, containerStatus := range newPod.Status.ContainerStatuses {
+		shortenContainerId := TruncateContainerId(containerStatus.ContainerID)
+		if shortenContainerId == "" {
+			continue
+		}
+		newContainerIds = append(newContainerIds, shortenContainerId)
+	}
+	containerIdCompare := compare.NewStringSlice(oldCachePod.ContainerIds, newContainerIds)
+	containerIdCompare.Compare()
+	deletedPodInfo.containerIds = containerIdCompare.GetRemovedElements()
+
+	// Check the ports specified.
+	newPorts := make([]int32, 0)
+	newHostPorts := make([]int32, 0)
+	for _, container := range newPod.Spec.Containers {
+		if len(container.Ports) == 0 {
+			newPorts = append(newPorts, 0)
+			continue
+		}
+		for _, port := range container.Ports {
+			newPorts = append(newPorts, port.ContainerPort)
+			// If hostPort is specified, add the container using HostIP and HostPort
+			if port.HostPort != 0 {
+				newHostPorts = append(newHostPorts, port.HostPort)
+			}
+		}
+	}
+
+	portsCompare := compare.NewInt32Slice(oldCachePod.Ports, newPorts)
+	portsCompare.Compare()
+	deletedPodInfo.ports = portsCompare.GetRemovedElements()
+
+	hostPortsCompare := compare.NewInt32Slice(oldCachePod.HostPorts, newHostPorts)
+	hostPortsCompare.Compare()
+	deletedPodInfo.hostPorts = hostPortsCompare.GetRemovedElements()
```

Comment:
When the pod IP changes but the port and hostPort remain unchanged, the hostIpPort and podIpPort information will not be updated normally.

But maybe the pod update could not modify the pod IP or pod Name.

Reply:
I have updated the testcase about this problem

Reply:
Actually, it could. I' ll fix this   issue.